lib/accesslists/visitor.go

```diff
@@ -0,0 +1,274 @@
+				return false, filteredOutReasonExpired
+			}
+			// If the membership is for a user but the not the one we are looking for, we filter it out.
+			if leg.member.Spec.MembershipKind == accesslist.MembershipKindUser && leg.member.Spec.Name != user.GetName() {
+				return false, ""
+			}
+		}
+
+		// If the member is a list but user doesn't meet the list's membership requirements, the leg is invalid.
+		if leg.list != nil && !UserMeetsRequirements(user, leg.list.Spec.MembershipRequires) {
+			return false, filteredOutReasonUnmetRequirements
+		}
+
+		return true, ""
+	}
+}
+
+// visitor visits all members of an AccessList graph by doing a depth-first traversal.
+// The visitor is cycle-proof.
+type visitor struct {
+	getter       AccessListAndMembersGetter
+	seen         map[string]struct{}
+	stack        []accessPath
+	filter       accessFilterFunc
+	ctx          context.Context
+	skippedPaths []skippedAccessPath
+}
+
+// accessPath represents a path in the access list graph from the start list to
+// a member.
+type accessPath []accessLeg
+
+// String implements stringer and provides a text representation of an accessPath.
+// This is used to explain access decisions in user-facing error messages.
+func (path accessPath) String() string {
+	var sb strings.Builder
+	for _, leg := range path {
+		if leg.member != nil {
+			sb.WriteString(" --> ")
+		}
+		if leg.list != nil {
+			sb.WriteString(leg.list.GetName())
+		} else {
+			sb.WriteString("user")
+		}
+	}
+	return sb.String()
+}
+
+// accessLeg represents one leg of an access path.
+// The first leg of the path has a nil member.
+// If the accessLeg target is an access list (as opposed to a user), list is non-nil.
+type accessLeg struct {
+	member *accesslist.AccessListMember
+	list   *accesslist.AccessList
+}
+
+// skippedAccessPath is an accessPath that got filtered out for a reason worth
+// surfacing to the end user/
+type skippedAccessPath struct {
+	accessPath
+	reason string
+}
+
+// newAccessPathVisitor returns a single-use iterator traversing the
+// nested access lists and returning access paths.
+// In case of non-nil error, the caller should stop processing as there's no
+// guarantee anymore that the graph will be completely traversed.
+// The caller can optionally pass a listFilterFunc to prevent the iterator from
+// visiting specific lists (e.g. restrict the graph traversal to lists a
+// specific user cam be member of).
+func newAccessPathVisitor(
+	ctx context.Context,
+	getter AccessListAndMembersGetter,
+	accessList *accesslist.AccessList,
+	filter accessFilterFunc) (*visitor, error) {
+
+	if filter == nil {
+		return nil, trace.BadParameter("filter is required (this is a bug)")
+	}
+
+	if accessList == nil {
+		return nil, trace.BadParameter("accessList is required (this is a bug)")
+	}
+
+	stack := make([]accessPath, 0)
+	skipped := make([]skippedAccessPath, 0)
+	start := accessLeg{
+		list: accessList,
+	}
+
+	ok, reason := filter(start)
```

Comment:
Nit: `ok, reason := filter(start)` could use a bit of "comprehensibility" enhancements.

Can we either rename the variables so they actually carry meaning (e.g. what is "ok"?) and/or add a comment explaining what we're filtering and why?

Reply:
I'd say this filter thing should be externalised. It feels a bit forced here for no reason (as far as I can tell). See https://github.com/gravitational/teleport/pull/60034#discussion_r2454936100